Add setup_context staticmethod to autograd.Function

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #90418
* #90415
* #90077
* #90240
* #90037
* #89860
* __->__ #89859
* #89858

Adds a setup_context staticmethod to autograd.Function.
If it exists, then the user splits the ctx-specific logic from the
forward() and puts it in the setup_context staticmethod.

Docs will come later when we remove the feature flag.

Test Plan:
- some light tests

zou3519 committed:
Add setup_context staticmethod to autograd.Function

Adds a setup_context staticmethod to autograd.Function.
If it exists, then the user splits the ctx-specific logic from the
forward() and puts it in the setup_context staticmethod.

Docs will come later when we remove the feature flag.

Test Plan:
- some light tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add setup_context staticmethod to autograd.Function"


Adds a setup_context staticmethod to autograd.Function.
If it exists, then the user splits the ctx-specific logic from the
forward() and puts it in the setup_context staticmethod.

Docs will come later when we remove the feature flag.

Test Plan:
- some light tests

cc ezyang albanD gqchen pearu nikitaved soulitzer Lezcano Varal7

[ghstack-poisoned]

soulitzer commented on test/test_autograd.py:
```diff
@@ -491,6 +500,91 @@ def fn(x):
+                return x.reshape(shape) * scale_forward
+
+            @staticmethod
+            def setup_context(ctx, outputs, x, shape, copy_forward, scale_backward):
```
Another option is to have both outputs and inputs non-expanded. Not sure how much better it is because users would always be able to just do `setup_context(ctx, outputs, *inputs)` anyway. It does seem more consistent and would allow inputs to be before outputs.

samdow replied:
> unexpanded inputs can be consistent across the various staticmethods

What happens right now if they aren't consistent? As an example, I could see a user trying to rewrite their existing autograd.Function and accidentally ending up with

```python
class Foo(torch.autograd.Function)
  @staticmethod
  def forward(ctx, x, y):
    ctx.save_for_backward(x, y) # a vestiage from the old way
    return torch.tensor(to_numpy(x) * to_numpy(y), device=x.device)

  @staticmethod
  def setup_context(ctx, outputs, x, y):
    ctx.save_for_backward(x, y)
```
and want to check what the error message looks like

zou3519 commented on test/test_autograd.py:
```diff
@@ -47,6 +47,15 @@
+        torch._C._set_autograd_function_extension_enabled(True)
+        yield
+    finally:
+        torch._C._set_autograd_function_extension_enabled(prev_state)
```
Sure. I'm going to use it a lot in the upcoming PRs

zou3519 replied:
Function moved.

zou3519 commented on test/test_autograd.py:
```diff
@@ -491,6 +500,91 @@ def fn(x):
+                return x.reshape(shape) * scale_forward
+
+            @staticmethod
+            def setup_context(ctx, outputs, x, shape, copy_forward, scale_backward):
```
Not a dumb question, this is a typo: it should be scale_forward

zou3519 replied:
Updated

samdow commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -877,29 +918,50 @@ PyObject* THPFunction_apply(PyObject* cls, PyObject* inputs) {
+      }
+      THPObjectPtr setup_context_fn(
+          PyObject_GetAttrString(cls, "setup_context"));
+      auto result =
+          PyObject_CallObject(setup_context_fn, ctx_output_input_tuple);
+      if (!result) {
+        return nullptr;
+      }
```
For my own understanding, when is result a nullptr? Mostly since result isn't used in the rest of this function so want to check that we aren't over aggressively returning nullptr

soulitzer replied:
Cpython API returns a nullptr when the call itself wasn't successful (I'm not sure when that actually happens, but I don't think we can recover from that), so we probably always want return nullptr ourselves too to propagate that error up again.

Related: https://docs.python.org/3/c-api/exceptions.html

zou3519 committed:
Update on "Add setup_context staticmethod to autograd.Function"


Adds a setup_context staticmethod to autograd.Function.
If it exists, then the user splits the ctx-specific logic from the
forward() and puts it in the setup_context staticmethod.

Docs will come later when we remove the feature flag.

Test Plan:
- some light tests

cc ezyang albanD gqchen pearu nikitaved soulitzer Lezcano Varal7

[ghstack-poisoned]

zou3519 committed:
Update on "Add setup_context staticmethod to autograd.Function"


Adds a setup_context staticmethod to autograd.Function.
If it exists, then the user splits the ctx-specific logic from the
forward() and puts it in the setup_context staticmethod.

Docs will come later when we remove the feature flag.

Test Plan:
- some light tests

cc ezyang albanD gqchen pearu nikitaved soulitzer Lezcano Varal7

[ghstack-poisoned]

zou3519 commented on test/test_autograd.py:
```diff
@@ -544,6 +544,91 @@ def fn(x):
+                return x.reshape(shape) * scale_forward
+
+            @staticmethod
+            def setup_context(ctx, outputs, x, shape, scale_forward, scale_backward):
```
Help wanted:
- should i call it `setup_ctx`?
- Are we OK with the inputs being expanded?

cc @albanD for more opinionated input

zou3519 committed:
Update on "Add setup_context staticmethod to autograd.Function"


Adds a setup_context staticmethod to autograd.Function.
If it exists, then the user splits the ctx-specific logic from the
forward() and puts it in the setup_context staticmethod.

Docs will come later when we remove the feature flag.

Test Plan:
- some light tests

cc ezyang albanD gqchen pearu nikitaved soulitzer Lezcano Varal7

[ghstack-poisoned]

zou3519 committed:
Update on "Add setup_context staticmethod to autograd.Function"


Adds a setup_context staticmethod to autograd.Function.
If it exists, then the user splits the ctx-specific logic from the
forward() and puts it in the setup_context staticmethod.

Docs will come later when we remove the feature flag.

Test Plan:
- some light tests

cc ezyang albanD gqchen pearu nikitaved soulitzer Lezcano Varal7

[ghstack-poisoned]

soulitzer reviewed (commented):
Looks good! But curious what Alban thinks about inputs being expanded.

zou3519 committed:
Update on "Add setup_context staticmethod to autograd.Function"


Adds a setup_context staticmethod to autograd.Function.
If it exists, then the user splits the ctx-specific logic from the
forward() and puts it in the setup_context staticmethod.

Docs will come later when we remove the feature flag.

Test Plan:
- some light tests

cc ezyang albanD gqchen pearu nikitaved soulitzer Lezcano Varal7

[ghstack-poisoned]

zou3519 committed:
Update on "Add setup_context staticmethod to autograd.Function"


Adds a setup_context staticmethod to autograd.Function.
If it exists, then the user splits the ctx-specific logic from the
forward() and puts it in the setup_context staticmethod.

Docs will come later when we remove the feature flag.

Test Plan:
- some light tests

cc ezyang albanD gqchen pearu nikitaved soulitzer Lezcano Varal7

[ghstack-poisoned]

zou3519 committed:
Update on "Add setup_context staticmethod to autograd.Function"


Adds a setup_context staticmethod to autograd.Function.
If it exists, then the user splits the ctx-specific logic from the
forward() and puts it in the setup_context staticmethod.

Docs will come later when we remove the feature flag.

Test Plan:
- some light tests

cc ezyang albanD gqchen pearu nikitaved soulitzer Lezcano Varal7

[ghstack-poisoned]

zou3519 committed:
Update on "Add setup_context staticmethod to autograd.Function"


Adds a setup_context staticmethod to autograd.Function.
If it exists, then the user splits the ctx-specific logic from the
forward() and puts it in the setup_context staticmethod.

Docs will come later when we remove the feature flag.

Test Plan:
- some light tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add setup_context staticmethod to autograd.Function"


Adds a setup_context staticmethod to autograd.Function.
If it exists, then the user splits the ctx-specific logic from the
forward() and puts it in the setup_context staticmethod.

Docs will come later when we remove the feature flag.

Test Plan:
- some light tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add setup_context staticmethod to autograd.Function"


Adds a setup_context staticmethod to autograd.Function.
If it exists, then the user splits the ctx-specific logic from the
forward() and puts it in the setup_context staticmethod.

Docs will come later when we remove the feature flag.

Test Plan:
- some light tests

[ghstack-poisoned]
